blockscout-ens/bens-logic/src/protocols/protocoler.rs

```diff
@@ -283,16 +292,100 @@ impl Protocoler {
         Ok(protocols)
     }
 
+    pub fn fetch_domain_options(
+        &self,
+        name_with_tld: &str,
+        network_id: i64,
+        maybe_filter: Option<NonEmpty<String>>,
+    ) -> Result<Vec<DomainNameOnProtocol>, ProtocolError> {
+        let tld = self.extract_tld(name_with_tld)?;
+
+        let protocols = self.protocols_of_network_for_tld(network_id, tld, maybe_filter)?;
+
+        let mut results = Vec::new();
+        for deployed_protocol in protocols {
+            let empty_label_hash = deployed_protocol
+                .protocol
+                .info
+                .protocol_specific
+                .empty_label_hash();
+
+            let domain_name = DomainName::new(name_with_tld, empty_label_hash)?;
+            results.push(DomainNameOnProtocol::from_str(
+                &domain_name.name,
+                deployed_protocol,
+            ));
+        }
+
+        let results = results
+            .into_iter()
+            .filter_map(|result| result.ok())
+            .collect::<Vec<_>>();
```

Comment:
why do you push result and after that check that result is ok.

you can simplify this to 
```
if let Ok(domain) = DomainNameOnProtocol::from_str() {
    results.push()
}
```

but after that, you can think about why do you need for-loop, Vec::new(), why do you use `DomainName::new` it's just useless, because you inside `DomainNameOnProtocol::from_str` you use `DomainName::new` again. after that your code will look like this:


```
let protocols = self.protocols_of_network_for_tld(network_id, tld, maybe_filter)?;
let domain_names = protocols
   .into_iter()
   .filter_map(|p| {
        DomainNameOnProtocol::from_str(name_with_tld, deployed_protocol).ok()
    })
if domain_names.is_empty() { blablabla };
Ok(domain_names)
```